### PR TITLE
[feat, UX] Add new OpenWithDialog based on RadioButton and RadioButtonTable

### LINKED ISSUE #3659: FR: Implement RadioButton and RadioButtonDialog
* KOReader version:  f6ca1c7 
* Device: all

#### Issue

Radio buttons can be quite useful.

### PR DESCRIPTION
**Pull request by Frenzie**
Fixes #3659 

screenshot_2018-02-11_16-56-08

### CODE REVIEW

**Comment by Frenzie** [comment]
As described in https://github.com/koreader/koreader/pull/3653#issuecomment-362377128

Pinging a few people for some thoughts @robert00s @poire-z @KenMaltby @Eduardomb22

File: frontend/document/documentregistry.lua [hunk 1]
```diff
@@ -179,15 +182,19 @@ function DocumentRegistry:showSetProviderButtons(file, filemanager_instance,  ui
...
                     })
-                end,
-            },
-        })
-        -- little trick for visual separation
-        table.insert(buttons, {})
-    end
+                end
+
+                -- just once
```
Comment (Frenzie) [inline frontend/document/documentregistry.lua]: Should probably be some kind of if elseif else construction instead. Potential lurking memory leak ?

Comment (poire-z) [reply frontend/document/documentregistry.lua]: Yes, it feels strange without any if/elseif/else, or no return in the if after the UIManager:show() :)

File: frontend/ui/widget/radiobutton.lua [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                    ges = "hold",
+                    range = self.dimen,
+                },
+                doc = "Hold Button",
+            }
+        }
+    end
+end
+
+function RadioButton:_initFrame()
```
Comment (poire-z) [inline frontend/ui/widget/radiobutton.lua]: I think most of the other widgets with this kind of 2 steps init, call this method `:update()`.
And although you don't need that as you don't have any subwidget that would really need it, you could add just there:
```lua
    if self[1] then
        self[1]:free()
    end
```

**Comment by poire-z** [comment]
Looking fine.
There could be more vertical space between the radio buttons and the checkboxes, or I dunno: it feels like the squared checkboxes apply only to the MuPDF entry. May be because of the line separator.

File: frontend/ui/widget/radiobutton.lua (hunk 2 above)
Comment (Frenzie) [reply frontend/ui/widget/radiobutton.lua]: True, might as well do that in case there's a BB or RichTextWidget in there in the future or something.

**Comment by Frenzie** [comment]
Alright, with a couple of tweaks:




@KenMaltby 

> User interaction looks good to me. So with neither "Always ..." checked it would be the one time thing? 

Yes, exactly. Same as the Microsoft Windows dialog. :-)

File: frontend/ui/widget/radiobutton.lua [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        return "partial", self.dimen
+    end)
+end
+
+function RadioButton:unCheck()
+    self._radio_button = self._unchecked_widget
+    self.checked = false
+    self:update()
+    UIManager:setDirty(self.parent, function()
+        return "partial", self.dimen
```
Comment (poire-z) [inline frontend/ui/widget/radiobutton.lua]: You could use `"ui"` for such a small widget (thinking about Kindle users with REAGL where `"partial"` is really a fast but noticable "full")

**Pull request merged**